functions/go/upsert-resource/upsertresource/upsert_resource.go

```diff
@@ -0,0 +1,167 @@
+	inputMeta, err := inputResource.GetMeta()
+	if err != nil {
+		return false, err
+	}
+	for i := range nodes {
+		rMeta, err := nodes[i].GetMeta()
+		if err != nil {
+			return false, err
+		}
+		// skip processing resource if it is a local config
+		if IsLocalConfig(rMeta) {
+			continue
+		}
+		// check if there is a match and replace the resource
+		if IsSameResource(inputMeta, rMeta) {
+			nodes[i], err = deepCopy(inputResource)
+			if err != nil {
+				return false, err
+			}
+			a := combineInputAndMatchedAnnotations(inputMeta.Annotations, rMeta.Annotations)
+			err = nodes[i].SetAnnotations(a)
+			if err != nil {
+				return false, err
+			}
+			// found a matching resource
+			found = true
+		}
+	}
+	return found, nil
+}
+
+// AddResource appends the inputResource to the list of input nodes
+// it also cleans up the meta annotations so that resource is created in new file
+// by the function orchestrator
+func AddResource(nodes []*yaml.RNode, inputResource *yaml.RNode) ([]*yaml.RNode, error) {
+	new, err := deepCopy(inputResource)
+	if err != nil {
+		return nodes, err
+	}
+	meta, err := new.GetMeta()
+	if err != nil {
+		return nodes, err
+	}
+	// remove local, function, path and index annotations from the result
+	// removing path and index annotations makes orchestrator write resource
+	// to a new file
+	cleanedAnno := removeMetaAnnotations(meta.Annotations)
+	err = new.SetAnnotations(cleanedAnno)
+	if err != nil {
+		return nodes, err
+	}
+	nodes = append(nodes, new)
+	return nodes, nil
+}
+
+// IsSameResource returns true if metadata of two resources
+// have same Group, Kind, Name, Namespace
+func IsSameResource(meta1, meta2 yaml.ResourceMeta) bool {
+	g1, _ := ParseGroupVersion(meta1.APIVersion)
+	g2, _ := ParseGroupVersion(meta1.APIVersion)
+	return g1 == g2 && meta1.Kind == meta2.Kind &&
+		meta1.Name == meta2.Name &&
+		meta1.Namespace == meta2.Namespace
+}
+
+// ParseGroupVersion parses a KRM metadata apiVersion field.
+func ParseGroupVersion(apiVersion string) (group, version string) {
+	if i := strings.Index(apiVersion, "/"); i > -1 {
+		return apiVersion[:i], apiVersion[i+1:]
+	}
+	return "", apiVersion
+}
+
+// combineInputAndMatchedAnnotations combines user provided non-meta annotations from inputResource,
+// with path and index annotations from matched resource and returns the result
+func combineInputAndMatchedAnnotations(inputResourceAnno, matchedResourceAnno map[string]string) map[string]string {
+	if inputResourceAnno == nil {
+		inputResourceAnno = make(map[string]string)
+	}
+	if matchedResourceAnno == nil {
+		matchedResourceAnno = make(map[string]string)
+	}
+	res := make(map[string]string)
+	// retain the annotations from the input resource in fn-config,
+	// these should be written to matched resource
+	for k, v := range inputResourceAnno {
+		res[k] = v
+	}
+	// retain the path and index annotation from matched resource to result
+	res[kioutil.PathAnnotation] = matchedResourceAnno[kioutil.PathAnnotation]
```

Comment:
I have seen 3 annotation added by kpt. We probably should retain all of them.
````
config.kubernetes.io/index: '0'
config.kubernetes.io/path: 'resources.yaml'
config.k8s.io/id: '2'
```

Reply:
We do not touch id annotation anywhere in the function. Updated the method name to be more specific.

Reply:
https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md

"A function MAY modify annotations with prefix config.kubernetes.io, but must be careful about doing so since they’re used for orchestration purposes and will likely impact subsequent functions in the pipeline."

Which means, don't just get rid of all config.kubernetes.io/* annotations. Instead retain all, and remove/update the ones you specifically care about.

Reply:
I am not,
If the resource is being added, I remove local, fn, path, index annotations.
If the resource is being replaced, I remove local, fn annotations from the fn-config resource and retain the path, index annotations of matched resource.

Updated method names and descriptions to be specific.